patroni/ha.py

```diff
@@ -46,6 +46,50 @@ def failover_limitation(self):
         return None
 
 
+class BackgroundKeepaliveSender(object):
+    """A context manager that sends keepalives every loop_wait seconds in a background thread while the context is
+    running, but only after a safepoint has been reached. After the safepoint PostgreSQL must not be allowed to
```

Comment:
And why we don't want to send keepalives before safepoint has been reached?

Also according to your comment from Oct 24, 2016 this code is mostly needed for `demote('offline')`:
> I did some research into how to reliably detect PostgreSQL shut down without having to wait for the shutdown checkpoint to complete. This is mostly relevant for demote('offline') case, as we need to hold off watchdog keepalives until PostgreSQL has completed shutdown process to the point that no backends survive, but having to wait for checkpointer to complete shutdown checkpoint would quite likely result in a reboot due to timeout.

But it seems that exactly for `demote('offline')` it is not used...